source/fuzz/transformation_add_copy_memory.cpp

```diff
@@ -0,0 +1,197 @@
+
+  // Check that |initializer_id| exists and has valid type.
+  const auto* initializer_inst =
+      ir_context->get_def_use_mgr()->GetDef(message_.initializer_id());
+  if (!initializer_inst || initializer_inst->type_id() != pointee_type_id) {
+    return false;
+  }
+
+  // Check that this transformation respects domination rules.
+  const auto* source_block = ir_context->get_instr_block(source_inst);
+  const auto* target_block = ir_context->get_instr_block(inst);
+  assert(target_block && "OpCopyMemory can't be a global instruction");
+
+  // Domination rules are satisfied if we are copying memory from a global
+  // instruction.
+  if (!source_block) {
+    return true;
+  }
+
+  // Check that both source and target instructions are in the same function.
+  if (source_block->GetParent() != target_block->GetParent()) {
+    return false;
+  }
+
+  // Check domination rules.
+  return source_inst != inst &&
+         ir_context->GetDominatorAnalysis(source_block->GetParent())
+             ->Dominates(source_inst, inst);
+}
+
+void TransformationAddCopyMemory::Apply(
+    opt::IRContext* ir_context,
+    TransformationContext* transformation_context) const {
+  // Insert OpCopyMemory before |instruction_descriptor|.
+  auto* insert_before_inst =
+      FindInstruction(message_.instruction_descriptor(), ir_context);
+  assert(insert_before_inst);
+
+  auto insert_before_iter = fuzzerutil::GetIteratorForInstruction(
+      ir_context->get_instr_block(insert_before_inst), insert_before_inst);
+
+  insert_before_iter.InsertBefore(MakeUnique<opt::Instruction>(
+      ir_context, SpvOpCopyMemory, 0, 0,
+      opt::Instruction::OperandList{
+          {SPV_OPERAND_TYPE_ID, {message_.fresh_id()}},
+          {SPV_OPERAND_TYPE_ID, {message_.source_id()}}}));
+
+  // Add global or local variable to copy memory into.
+  auto storage_class = static_cast<SpvStorageClass>(message_.storage_class());
+  auto type_id = fuzzerutil::MaybeGetPointerType(
+      ir_context,
+      fuzzerutil::GetPointeeTypeIdFromPointerType(
+          ir_context, fuzzerutil::GetTypeId(ir_context, message_.source_id())),
+      storage_class);
+
+  if (storage_class == SpvStorageClassPrivate) {
+    fuzzerutil::AddGlobalVariable(ir_context, message_.fresh_id(), type_id,
+                                  storage_class, message_.initializer_id());
+  } else {
+    assert(storage_class == SpvStorageClassFunction &&
+           "Storage class can be either Private or Function");
+    fuzzerutil::AddLocalVariable(ir_context, message_.fresh_id(), type_id,
+                                 ir_context->get_instr_block(insert_before_inst)
+                                     ->GetParent()
+                                     ->result_id(),
+                                 message_.initializer_id());
+  }
+
+  fuzzerutil::UpdateModuleIdBound(ir_context, message_.fresh_id());
+
+  transformation_context->GetFactManager()->AddFactValueOfPointeeIsIrrelevant(
+      message_.fresh_id());
```

Comment:
I think it would be good to have a brief comment before this, along the lines of:

"Even though the copy memory instruction will - at least temporarily - lead to the destination and source pointers referring to identical values, this fact is not guaranteed to hold throughout execution of the SPIR-V code since the source pointer could be over-written.  We thus assume nothing about the destination pointer, and record this fact so that the destination pointer can be used freely by other fuzzer passes."

Reply:
Fixed in d32f2fd.